frappe/database/postgres/schema.py

```diff
@@ -73,37 +88,54 @@ def alter(self):
 
 			query.append("ALTER COLUMN `{}` SET DEFAULT {}".format(col.fieldname, col_default))
 
-		create_index_query = ""
+		create_contraint_query = ""
 		for col in self.add_index:
 			# if index key not exists
-			create_index_query += 'CREATE INDEX IF NOT EXISTS "{index_name}" ON `{table_name}`(`{field}`);'.format(
+			create_contraint_query += 'CREATE INDEX IF NOT EXISTS "{index_name}" ON `{table_name}`(`{field}`);'.format(
 				index_name=col.fieldname,
 				table_name=self.table_name,
 				field=col.fieldname)
 
-		drop_index_query = ""
+		for col in self.add_unique:
+			# if index key not exists
+			create_contraint_query += 'CREATE UNIQUE INDEX IF NOT EXISTS "unique_{index_name}" ON `{table_name}`(`{field}`);'.format(
+				index_name=col.fieldname,
+				table_name=self.table_name,
+				field=col.fieldname
+			)
+
+		drop_contraint_query = ""
 		for col in self.drop_index:
 			# primary key
 			if col.fieldname != 'name':
 				# if index key exists
-				if not frappe.db.has_index(self.table_name, col.fieldname):
-					drop_index_query += 'DROP INDEX IF EXISTS "{}" ;'.format(col.fieldname)
+				drop_contraint_query += 'DROP INDEX IF EXISTS "{}" ;'.format(col.fieldname)
 
-		if query:
-			try:
+		for col in self.drop_unique:
+			# primary key
+			if col.fieldname != 'name':
+				# if index key exists
+				drop_contraint_query += 'DROP INDEX IF EXISTS "unique_{}" ;'.format(col.fieldname)
+		try:
+			if query:
 				final_alter_query = "ALTER TABLE `{}` {}".format(self.table_name, ", ".join(query))
-				if final_alter_query: frappe.db.sql(final_alter_query)
-				if create_index_query: frappe.db.sql(create_index_query)
-				if drop_index_query: frappe.db.sql(drop_index_query)
-			except Exception as e:
-				# sanitize
-				if frappe.db.is_duplicate_fieldname(e):
-					frappe.throw(str(e))
-				elif frappe.db.is_duplicate_entry(e):
-					fieldname = str(e).split("'")[-2]
-					frappe.throw(_("""{0} field cannot be set as unique in {1},
-						as there are non-unique existing values""".format(
-						fieldname, self.table_name)))
-					raise e
-				else:
-					raise e
+				# nosemgrep
+				frappe.db.sql(final_alter_query)
+			if create_contraint_query:
+				# nosemgrep
+				frappe.db.sql(create_contraint_query)
+			if drop_contraint_query:
+				# nosemgrep
+				frappe.db.sql(drop_contraint_query)
+		except Exception as e:
+			# sanitize
+			if frappe.db.is_duplicate_fieldname(e):
+				frappe.throw(str(e))
+			elif frappe.db.is_duplicate_entry(e):
+				fieldname = str(e).split("'")[-2]
+				frappe.throw(
```

Comment:
from what I have heard, this constraint breaks frequently - so better to add a suffix like `- 1` etc to the uniques. In most cases this should be acceptable

Reply:
This failure does not occur while adding a row (in which case adding `-1` was possible)... instead, this arises when we add the "Unique" constraint to a column with existing data with multiple not unique values. In this case, a user might not expect the system to update multiple values of a column (with some prefix).